Click order fix

oetiker opened the pull request:
Fix for #10631 

```javascript
// Document is the application root
let evCnt = 0;
let doc = this.getRoot();
let tf = new qx.ui.form.TextField('change me');
doc.add(tf, { left:5, top: 5});
tf.addListener('changeValue',(e) => { evCnt ++; tf.debug(`field changeValue -> ${evCnt}`)});
let bt = new qx.ui.form.Button("then click");
doc.add(bt, { left:5, top: 40})
// without this, the event order on a touch interface will be
// not as expected ...
bt.addListenerOnce('appear',() => {
  let el = this.getContentElement().getDomElement();
  bt.addListener('touchstart', () => { el.focus();  });
});
bt.addListener('execute',(e) => { evCnt++; bt.debug(`button click -> ${evCnt}`)});
```

PLEASE test on your mobile device!

oetiker committed:
Update Button.js

It seems that touching a button (at least on ios and android)
does not immediately focus it before triggering the tap event
this causes problem with change events for input fields that
previously held focus, as they will arrive AFTER the button has
already executed.

oetiker committed:
Update AbstractField.js

This is not necessary anymore as the focus changes early enough

johnspackman commented:
> PLEASE test on your mobile device!

Works on iOS 16.6.1

hkollmann reviewed (commented):
Sure you want to delete the code you just inserted a few days ago?

oetiker commented:
> Sure you want to delete the code you just inserted a few days ago?

yes :) I am pretty sure this is a better solution to the problem

oetiker commented:
Will do some validation testing today in my application

hkollmann committed:
Merge branch 'master' into click-order-fix

hkollmann committed:
Merge branch 'master' into click-order-fix

oetiker commented:
the fix is not complete ... it seems when clicking very quickly the event order still is reversed.

goldim commented:
@oetiker should I test it on device now?

oetiker committed:
Update Button.js

switch to using native events

oetiker commented:
I think now I found a working solution.

oetiker committed:
Update Button.js

the event handler is fine ... the problem is just the 'focus' call .. that needs to be on the DOM element

oetiker commented on source/class/qx/ui/form/Button.js:
```diff
@@ -80,11 +80,17 @@ qx.Class.define("qx.ui.form.Button", {
+    // this causes problem with change events for input fields that
+    // previously held focus, as they will arive AFTER the button has
+    // already executed.
+    this.addListener("touchstart", () => { this.focus() }, this);
```
this happens not only on android ... on ios too, that is why I removed my previous PR ... also it seems to happen on chrome and on firefox on android ... but adding it on the widget level might make sense ... and protect it with 'isFocusable'

oetiker replied:
Thinking on it a bit more – it is fine when this is only done on the button widget … because the event ordering is only a problem in this context.

goldim reviewed (approved):
@oetiker Tested your examples from issue and this PR (btw this code causes an error in console). It works the same on android (old honor 10) and my pc. Tbh I am a little scared of applying to all widgets (not mobile devices only) but if everyone is fine with it then let it go.

oetiker committed:
Update source/class/qx/ui/form/Button.js

Co-authored-by: Dmitry Zolotov <goldim1993@mail.ru>

oetiker commented:
> @oetiker Tested your examples from issue and this PR (btw this code causes an error in console). It works the same on android (old honor 10) and my pc. Tbh I am a little scared of applying to all widgets (not mobile devices only) but if everyone is fine with it then let it go.

Note that the code only acts on touchstart, so normal desktop (mouse) operation will not be affected ... it would be interesting to know what happens when this is tried on a touch laptop ...

hkollmann committed:
Merge branch 'master' into click-order-fix

oetiker commented:
@level420 and @derrell any comments ?
